flang/lib/Lower/ReductionRuntime.cpp

```diff
@@ -714,6 +798,84 @@ mlir::Value Fortran::lower::genProduct(Fortran::lower::FirOpBuilder &builder,
+  else if (eleTy == fir::ComplexType::get(builder.getContext(), 4))
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(CppDotProductComplex4)>(
+        loc, builder);
+  else if (eleTy == fir::ComplexType::get(builder.getContext(), 8))
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(CppDotProductComplex8)>(
+        loc, builder);
+  else if (eleTy == fir::ComplexType::get(builder.getContext(), 10))
+    func =
+        Fortran::lower::getRuntimeFunc<ForcedDotProductComplex10>(loc, builder);
+  else if (eleTy == fir::ComplexType::get(builder.getContext(), 16))
+    func =
+        Fortran::lower::getRuntimeFunc<ForcedDotProductComplex16>(loc, builder);
+  else if (eleTy ==
+           builder.getIntegerType(builder.getKindMap().getIntegerBitsize(1)))
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(DotProductInteger1)>(loc,
+                                                                       builder);
+  else if (eleTy ==
+           builder.getIntegerType(builder.getKindMap().getIntegerBitsize(2)))
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(DotProductInteger2)>(loc,
+                                                                       builder);
+  else if (eleTy ==
+           builder.getIntegerType(builder.getKindMap().getIntegerBitsize(4)))
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(DotProductInteger4)>(loc,
+                                                                       builder);
+  else if (eleTy ==
+           builder.getIntegerType(builder.getKindMap().getIntegerBitsize(8)))
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(DotProductInteger8)>(loc,
+                                                                       builder);
+  else if (eleTy ==
+           builder.getIntegerType(builder.getKindMap().getIntegerBitsize(16)))
+    func =
+        Fortran::lower::getRuntimeFunc<ForcedDotProductInteger16>(loc, builder);
+  else if (eleTy.isa<fir::LogicalType>())
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(DotProductLogical)>(loc,
+                                                                      builder);
+  else
+    fir::emitFatalError(loc, "invalid type in DotProduct lowering");
+
+  auto fTy = func.getType();
+  auto sourceFile = Fortran::lower::locationToFilename(builder, loc);
+
+  if (fir::isa_complex(eleTy)) {
+    auto sourceLine =
+        Fortran::lower::locationToLineNo(builder, loc, fTy.getInput(4));
+    auto args = Fortran::lower::createArguments(builder, loc, fTy, resultBox,
+                                                vectorABox, vectorBBox,
+                                                sourceFile, sourceLine);
+    builder.create<fir::CallOp>(loc, func, args);
+    return resultBox;
+  }
+
+  auto sourceLine =
+      Fortran::lower::locationToLineNo(builder, loc, fTy.getInput(3));
+  auto args = Fortran::lower::createArguments(
+      builder, loc, fTy, vectorABox, vectorBBox, sourceFile, sourceLine);
```

Comment:
I see the same in genSum and genProd, but why is the resultBox not included here?

Reply:
In this case, it's similar for `product`: https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/ReductionRuntime.cpp#L712-L713. `resultBox` **is** included when `if (fir::isa_complex(eleTy))` returns `true` (L863).

Reply:
Correct. resultBox is only used with the complex case in these functions.